File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:

        res = nums[0]
        curMin , curMax = nums[0] , nums[0]

        for n in nums[1:]:
            temp = curMax;
            curMax = max(n , temp*n , curMin*n)
            curMin = min(n , temp*n , curMin*n)
            res = max(res,curMax)
        
        return res

        

File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:

    int solve(int i, vector<int>& nums ,  vector<int>& dp){
            int n = nums.size();
            if(i>=n){
                return 0;
            }
            if(dp[i] != -1){
                return dp[i];
            }
            int take = nums[i] + solve(i+2 , nums , dp);
            int skip = solve(i+1 , nums ,dp ) ;
            dp[i] =max(take,skip);
            return dp[i];
        
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n+1 , -1);

        return solve(0 , nums , dp);
        
    }
};
